cli/internal/runsummary/spaces.go

```diff
@@ -1,13 +1,235 @@
 package runsummary
 
 import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/mitchellh/cli"
 	"github.com/vercel/turbo/cli/internal/ci"
+	"github.com/vercel/turbo/cli/internal/client"
 )
 
-// spacesRunResponse deserialized the response from POST Run endpoint
-type spacesRunResponse struct {
-	ID  string
-	URL string
+const runsEndpoint = "/v0/spaces/%s/runs"
+const runsPatchEndpoint = "/v0/spaces/%s/runs/%s"
+const tasksEndpoint = "/v0/spaces/%s/runs/%s/tasks"
+
+// spaceRequest contains all the information for a single request to Spaces
+type spaceRequest struct {
+	method  string
+	url     string
+	body    interface{}
+	makeURL func(self *spaceRequest, r *spaceRun) error // Should set url on self
+	onDone  func(self *spaceRequest, response []byte)   // Handler for when request completes
+}
+
+func (req *spaceRequest) error(msg string) error {
+	return fmt.Errorf("[%s] %s: %s", req.method, req.url, msg)
+}
+
+type spacesClient struct {
+	rsm      *Meta
+	requests chan *spaceRequest
+	errors   []error
+	api      *client.APIClient
+	ui       cli.Ui
+	run      *spaceRun
+	wg       sync.WaitGroup
+	spaceID  string
+	enabled  bool
+}
+
+type spaceRun struct {
+	ID      string
+	URL     string
+	created chan struct{} // a signal that the run has completed
+}
+
+func newSpacesClient(spaceID string, api *client.APIClient, ui cli.Ui, rsm *Meta) *spacesClient {
+	return &spacesClient{
+		api:      api,
+		ui:       ui,
+		rsm:      rsm,
+		spaceID:  spaceID,
+		enabled:  spaceID != "",
+		requests: make(chan *spaceRequest), // TODO: give this a size based on tasks
+		// Set a default, empty one here, so we'll have something downstream and not a segfault
+		run: &spaceRun{
+			created: make(chan struct{}, 1),
+		},
+	}
+}
+
+// Start receiving and processing requests in 8 goroutines
+// There is an additional marker (protected by a mutex) that indicates
+// when the first request is done. All other requests are blocked on that one.
+// This first request is the POST /run request. We need to block on it because
+// the response contains the run ID from the server, which we need to construct the
+// URLs of subsequent requests.
+func (c *spacesClient) start() {
+	if !c.enabled {
+		return
+	}
+
+	mu := sync.Mutex{}
+	firstReqDone := false
+	processors := 8
+	for i := 0; i < processors; i++ {
+		c.wg.Add(1)
+		go func() {
+			defer c.wg.Done()
+			for req := range c.requests {
+				// since we have multiple processors, we want to lock firstReqDone so only the first one will mark firstReqDone
+				mu.Lock()
+				if !firstReqDone {
+					firstReqDone = true
+					mu.Unlock()
+					go func(r *spaceRequest) {
+						c.makeRequest(r)
+						close(c.run.created) // close this channel to signal that other requests can proceed
+					}(req)
+
+				} else {
+					mu.Unlock()
+
+					// If this is not the first request, wait for the run to be created
+					<-c.run.created
+					go c.makeRequest(req)
```

Comment:
I'm not I understand the intention here. Are you intending to limit to 8 concurrent requests? This line starts another goroutine to do the request, meaning you will have 8 goroutines reading from the requests channel and then spinning up a goroutine for each request. 

A goroutine-per-request is a totally reasonable Go thing to do, it's just weird to have 8 goroutines to kick them off.

If you intend to have 8 concurrent requests, drop the `go` statement here. 
If instead you intend to have a goroutine per request, I would have a single goroutine read from the channel. If the first request hasn't finished yet, buffer the requests. Once the first request finishes, fire off everything in the buffer and directly fire off anything else that comes from the channel.

Reply:
I started with this pattern:
https://github.com/vercel/turbo/blob/c1f3e666db38f466ae5d0f2008f76b3618260069/cli/internal/run/dry_run.go#L109-L111
and
https://github.com/vercel/turbo/blob/c1f3e666db38f466ae5d0f2008f76b3618260069/cli/internal/cache/async_cache.go#L36-L38

but after setting up 8 workers to process requests from the channel, I observed that once they are unblocked by the first request, they seem to become blocking code and block task execution. So I added another goroutine-per-request. 

Conversely, if I only made a goroutine per-request, I also observed some congestion because each request blocked the next one.

It's entirely possible I did something in each of these approaches, but this double go-routine approach is how I was able to land with the total execution time of my test request to actually be the same as without making these requests in the first place.

Reply:
@gsoltis I took another stab at this here, mind taking another look? https://github.com/vercel/turbo/pull/4924